code/modules/underworld/underworld.dm

```diff
@@ -39,6 +39,130 @@
+
+#define CANT_LEAVE_ITEMS \
+	/obj/item/rogueweapon/woodstaff/aries, \
+	/obj/item/clothing/head/roguetown/crown, \
+	/obj/item/key/lord
+
+/obj/structure/carriagedoor
+	name = "Carriage door"
+	desc = "The wooden entrance to the carriage. It allows those who come to arrive and those who leave to depart."
+	icon = 'icons/roguetown/misc/doors.dmi'
+	icon_state = "carriage_door"
+	layer = WALL_OBJ_LAYER
+	plane = GAME_PLANE_UPPER
+	pixel_y = 27
+	max_integrity = 9999 // In case any maniac decides to hit it.
+
+
+/obj/structure/carriagedoor/attack_hand(mob/user)
+	. = ..()
+
+	var/mob/living/carbon/human/H
+	if(ishuman(user))
+		H = user
+
+	if(!isliving(user) || user.incapacitated())
+		return //No ghosts or incapacitated folk allowed to do this.
+
+	if(user.get_active_held_item())
+		to_chat(user, span_warn("I cant turn the handle. My hand has something in it!"))
+		return
+
+	var/datum/job/J = SSjob.name_occupations[user.job]
+	for(var/K in list(CANT_LEAVE_ROLES))
+		if(istype(J, K))
+			to_chat(user, span_warning("You are too important to leave!!"))
+			return
+
+	// No important antags can leave!
+	if(H.mind && H.mind.antag_datums)
+		var/datum/mind/M = H.mind
+		for(var/a in M.antag_datums)
+			for(var/b in ALL_ANTAG_TYPES)
+				if(istype(a, b)) 
+					to_chat(user, span_warning("I still have work to do here..."))
+					return
+
+	//This check tree is the face of hell. If someone knows how to clean this up please help.
+	//Get everything (YES EVEN INSIDE ITEMS, HANDS TOO!)
+	var/list/user_items = H.get_equipped_items()
+	user_items += H.held_items 
+	for(var/obj/item/i in user_items)
+		for(var/j in list(CANT_LEAVE_ITEMS))
+			if(istype(i, j))
+				to_chat(user, span_notice("You can't leave with \the [i.name]!!"))
+				return
+				
+		if(SEND_SIGNAL(i, COMSIG_CONTAINS_STORAGE))
+			var/list/inv = list()
+
+			SEND_SIGNAL(i, COMSIG_TRY_STORAGE_RETURN_INVENTORY, inv, FALSE)
+
+			for(var/obj/item/L in inv)
+				for(var/M in list(CANT_LEAVE_ITEMS))
+					if(istype(L, M))
+						to_chat(user, span_notice("You can't leave with \the [L.name]!!"))
+						return
+
+
+	switch(alert("Do you wish to leave town? (You cannot return.)",,"Yes","No"))
+		if("Yes")
+			to_chat(user, span_notice("You start climbing into the carriage to leave..."))
+			if(do_after(user, 5 SECONDS))
+				// Thank you Azure <3 I did not want to have to write this.
+				// Logs everything on the mob
+				var/dat = "[key_name(user)] has departed town via the carriage. job [user.job], at [AREACOORD(src)]. They had:"
+				if(!length(user.contents))
+					dat += " Nothing on them."
+				else
+					// WARNING: This doesn't currently list items in the storage
+					// But since we checked for shit that matters before already do we really care?
+					var/atom/movable/content = user.contents[1]
+					dat += " [content.name]"
+					for(var/i in 2 to length(user.contents))
+						content = user.contents[i]
+						dat += ", [content.name]"
+						dat += "."
```

Comment:
```suggestion
				var/list/contained_items = user.get_all_gear()
				if(!length(contained_items))
					dat += " Nothing on them."
				else
					dat += " [english_list(user.get_all_gear())]"
```
This will list all items including those in storage.